### DISCUSSION TITLE
Add an Anonymization middleware - scrub request and return context after receiving response

### DISCUSSION THREAD

**Discussion by karayanni** [open-webui/open-webui#3222] [Ideas]
Privacy and Security are critical, with the proven strength of Larger models - running local models is limiting capabilities.
Similar to what Apple announced at the WWDC24 - the hybrid approach is gaining traction. where a combination of local models work with the possibility of sending to a server LLM if needed.

While apple gave many guarantees about their cloud's security. it is still a significant attack surface and risk factor.

We propose a feature to add a new layer of anonymization before sending to the server LLM. similar to LLM-Guard's features.

**Comment by matthid** [comment]
I thought about something similar: Why not use a local model to anonymize the request (business internals):
 microsoft.com -> company.com
On the way in and (maybe as a future step) translate it back when the request from the cloud provider returns. Some people will strip confidential information by hand at the moment, but it would be nice to have a 'safe' input box and a preview and if it is anonymized enough you can send it.

Would a PR for this be accepted? I think this would be quite useful for companies.

**Reply by rgaricano** [reply]
It can be done with a function.

**Reply by matthid** [reply]
Well it depends, can a function change the UI? I didn't think this should be fully automatic, I'd expect the user to verify if it is properly anonymized, potentially even multiple follow ups until it is ready to be sent to the cloud
